cirrus-ci_env/cirrus-ci_env.py

```diff
@@ -0,0 +1,259 @@
+        return out
+
+    @staticmethod
+    def format_env(env, global_env: dict) -> dict:
+        """Replace shell-style references in env values, from global_env then env."""
+        # This method is also used to initialize self.global_env
+        if global_env is None:
+            global_env = dict()
+
+        rep = r"{\1}"  # Shell env var to python format string conversion regex
+        def_fmt = DefFmt(**global_env)  # Assumes global_env already rendered
+
+        for k, v in env.items():
+            if "ENCRYPTED" in str(v):
+                continue
+            _ = def_fmt.dollarcurly_env_var.sub(rep, str(v))
+            def_fmt[k] = def_fmt.dollar_env_var.sub(rep, _)
+        out = dict()
+        for k, v in def_fmt.items():
+            if k in env:  # Don't unnecessarily duplicate globals
+                out[k] = str(v).format_map(def_fmt)
+        return out
+
+    def render_tasks(self, tasks: dict) -> dict:
+        """Return new tasks dict with envs rendered and matrices unrolled."""
+        result = {}
+        for k, v in tasks.items():
+            if not k.endswith("_task"):
+                continue
+            # Cirrus-CI uses this defaulting priority order
+            alias = v.get("alias", k.replace("_task", ""))
+            name = v.get("name", alias)
+            if "matrix" in v:
+                # Assume Cirrus-CI accepted this config., don't check name clashes
+                result.update(self.unroll_matrix(name, alias, v))
+            else:
+                task = dict(alias=alias)
+                task["env"] = self.render_env(v.get("env", dict()))
+                task_name = self.render_value(name, task["env"])
+                _ = self.get_type_image(v, self.global_type, self.global_image)
+                self.init_task_type_image(task, *_)
+                result[task_name] = task
+        return result
+
+    def unroll_matrix(self, name_default: str, alias_default: str, task: dict) -> dict:
+        """Produce copies of task with attributes replaced from matrix list."""
+        result = {}
+        for item in task["matrix"]:
+            if "name" not in task and "name" not in item:
+                # Cirrus-CI goes a step further, attempting to generate a
+                # unique name based on alias + matrix attributes.  This is
+                # a very complex process that would be insane to attempt to
+                # duplicate.  Instead, simply require a defined 'name'
+                # attribute in every case, throwing an error if not found.
+                raise ValueError(f"Expecting 'name' attribute in"
+                                 f" '{alias_default}_task'"
+                                 f" or matrix definition: {item}"
+                                 f" for task definition: {task}")
+            # default values for the rendered task
+            matrix_task = dict(alias=alias_default, env=task.get("env").copy())
```

Comment:
```suggestion
            matrix_task = dict(alias=alias_default, env=task.get("env"))
```
If you are already creating a new dict() do you need to copy the task["env"] one? I guess it depends on whether task["env"] is mutable. If it is then the method function should use the type MutableMapping[] rather than dict.

Reply:
> I guess it depends on whether task["env"] is mutable.

No it must *not* be mutable, in fact the unit-tests specifically check for this..  I should make this clear with a comment.